astropy/uncertainty/core.py

```diff
@@ -43,40 +58,143 @@ def __new__(cls, samples):
+        # Set our new structured dtype.
+        structured.dtype = new_dtype
+        # Get rid of trailing dimension of 1.
+        structured.shape = samples.shape[:-1]
+
+        # Now view as the Distribution subclass, and finalize based on the
+        # original samples (e.g., to set the unit for QuantityDistribution).
+        new_cls = cls._get_distribution_cls(type(samples))
+        self = structured.view(new_cls)
+        if not NUMPY_LT_1_23 or callable(self.__array_finalize__):
+            self.__array_finalize__(samples)
+        return self
+
+    @classmethod
+    def _get_distribution_cls(cls, samples_cls):
+        if issubclass(samples_cls, Distribution):
+            return samples_cls
+
         new_cls = cls._generated_subclasses.get(samples_cls)
         if new_cls is None:
-            # Make a new class with the combined name, inserting Distribution
-            # itself below the samples class since that way Quantity methods
-            # like ".to" just work (as .view() gets intercepted).  However,
-            # repr and str are problems, so we put those on top.
-            # TODO: try to deal with this at the lower level.  The problem is
-            # that array2string does not allow one to override how structured
-            # arrays are typeset, leading to all samples to be shown.  It may
-            # be possible to hack oneself out by temporarily becoming a void.
-            new_name = samples_cls.__name__ + cls.__name__
+            # Walk through MRO and find closest base data class.
+            # Note: right now, will basically always be ndarray, but
+            # one could imagine needing some special care for one subclass,
+            # which would then get its own entry.  E.g., if MaskedAngle
+            # defined something special, then MaskedLongitude should depend
+            # on it.
+            for mro_item in samples_cls.__mro__:
+                base_cls = cls._generated_subclasses.get(mro_item)
+                if base_cls is not None:
+                    break
+            else:
+                # Just hope that NdarrayDistribution can handle it.
+                # TODO: this covers the case where a user puts in a list or so,
+                # but for those one could just explicitly do something like
+                # _generated_subclasses[list] = NdarrayDistribution
+                return NdarrayDistribution
+
+            # We need to get rid of the top _DistributionRepr, since we add
+            # it again below (it should always be on top).
+            # TODO: remove the need for _DistributionRepr by defining
+            # __array_function__ and overriding array2string.
+            if base_cls.__mro__[1] is not _DistributionRepr:
+                # Sanity check. This should never happen!
+                raise RuntimeError(
+                    f"found {base_cls=}, which does not have _DistributionRepr at "
+                    "__mro__[1]. Please raise an issue."
```

Comment:
```suggestion
                    "__mro__[1]. Please raise an issue at "
                    "https://github.com/astropy/astropy/issues/new/choose."
```

Reply:
Yes, I should not have been that lazy... Now corrected.